src/core/src/bound_evaluate.cpp

```diff
@@ -75,85 +75,219 @@ bool is_type_allocable(const element::Type& type) {
     return type != element::undefined && type.is_static();
 }
 
-ov::Tensor evaluate_bound(const Output<Node>& output, bool is_upper, bool invalidate_all_unused_values = true) {
-    if (is_upper && output.get_tensor().get_upper_value()) {
-        return output.get_tensor().get_upper_value();
-    }
+/**
+ * @brief Checks if node's inputs has bound set and are equal.
+ *
+ * @param node  Node
+ * @return true if all inputs have both bounds set and are equal, otherwise false.
+ */
+bool node_has_same_inputs(const Node& node) {
+    const auto& values = node.input_values();
+
+    return std::all_of(values.begin(), values.end(), [](const Output<Node>& value) {
+        const auto& t_desc = value.get_tensor();
+        return t_desc.has_and_set_bound() || are_equal(t_desc.get_lower_value(), t_desc.get_upper_value());
+    });
+};
 
-    if (!is_upper && output.get_tensor().get_lower_value()) {
-        return output.get_tensor().get_lower_value();
+/**
+ * @brief Invalidates unused tensor values for outputs (except Constants)
+ *
+ * @param outputs  Outputs to apply invalidation.
+ */
+void invalidate_unused_values(const ov::OutputVector& outputs) {
+    for (const auto& output : outputs) {
+        auto& tensor = output.get_tensor();
+        const auto& lower = tensor.get_lower_value();
+        const auto& upper = tensor.get_upper_value();
+        const auto should_invalidate =
+            (lower && shape_size(lower.get_shape()) > 10) || (upper && shape_size(upper.get_shape()) > 10);
+        if (should_invalidate && output.get_target_inputs().size() == 1) {
+            tensor.invalidate_values();
+        }
     }
+}
 
-    std::vector<Node*> order;
-    if (could_propagate(output, order)) {
-        for (const auto& node : order) {
-            ov::TensorVector outputs;
-            for (const auto& out : node->outputs()) {
-                if (is_type_allocable(out.get_element_type())) {
-                    outputs.emplace_back(out);
-                } else {
-                    outputs.emplace_back();
-                }
+/**
+ * @brief Process and remove nodes with bounds
+ *
+ * @param nodes  Vector of nodes pointers (no nullptr) to processed modified.
+ * @param output Destination of RT info  propagation from processed nodes.
+ */
+static void process_and_rm_nodes_with_bounds(std::vector<Node*>& nodes, const Output<Node>& output) {
+    auto first_has_bounds = std::remove_if(nodes.begin(), nodes.end(), [&output](Node* n) {
+        if (is_type<op::v0::Constant>(n)) {
+            propagate_rt_info(n, output);
+            return true;
+        } else {
+            return false;
+        }
+    });
+    nodes.erase(std::move(first_has_bounds), nodes.end());
```

Comment:
Does that `std::move` have a better performance than just const ref?

Reply:
Iterators are passed by value. The move can have no effect depends what kind of object is iterator but will should prevent Coverity complains about make a copy instead move as `first_has_bounds` not required any more.